File: ios/FlowerLines/Rendering/GridRenderer.swift
import UIKit

struct GridRenderer {
    let layout: Layout
    let state: GameState

    // MARK: - Grid Background
    func drawGrid(ctx: CGContext) {
        let L = layout
        // Green gradient background
        let grad = CGGradient(
            colorsSpace: CGColorSpaceCreateDeviceRGB(),
            colors: [UIColor(hex: "#dcedc8").cgColor, UIColor(hex: "#c5e1a5").cgColor] as CFArray,
            locations: [0, 1])!
        ctx.drawLinearGradient(grad,
            start: CGPoint(x: 0, y: L.headerHeight),
            end:   CGPoint(x: 0, y: L.headerHeight + L.cellSize * CGFloat(GRID_SIZE)),
            options: [])

        // Checkerboard + grid lines
        for r in 0..<GRID_SIZE {
            for c in 0..<GRID_SIZE {
                let x = CGFloat(c) * L.cellSize
                let y = L.headerHeight + CGFloat(r) * L.cellSize
                let rect = CGRect(x: x, y: y, width: L.cellSize, height: L.cellSize)
                if (r + c) % 2 == 0 {
                    ctx.setFillColor(UIColor(white: 1, alpha: 0.22).cgColor)
                    ctx.fill(rect)
                }
                ctx.setStrokeColor(UIColor(red: 0.51, green: 0.71, blue: 0.47, alpha: 0.55).cgColor)
                ctx.setLineWidth(0.7)
                ctx.stroke(rect.insetBy(dx: 0.35, dy: 0.35))
            }
        }
    }

    // MARK: - Highlights
    func drawHighlights(ctx: CGContext) {
        guard let moves = state.validMoves else { return }
        let L = layout
        for pos in moves {
            let x = CGFloat(pos.col) * L.cellSize + 1
            let y = L.headerHeight + CGFloat(pos.row) * L.cellSize + 1
            ctx.setFillColor(UIColor(red: 1, green: 0.9, blue: 0.2, alpha: 0.40).cgColor)
            ctx.fill(CGRect(x: x, y: y, width: L.cellSize-2, height: L.cellSize-2))
            ctx.setFillColor(UIColor(red: 0.78, green: 0.67, blue: 0, alpha: 0.25).cgColor)
            ctx.addEllipse(in: CGRect(x: L.cellCX(pos.col)-4, y: L.cellCY(pos.row)-4, width: 8, height: 8))
            ctx.fillPath()
        }
    }

    // MARK: - Score Bar
    func drawScoreBar(ctx: CGContext) {
        let L = layout
        let grad = CGGradient(
            colorsSpace: CGColorSpaceCreateDeviceRGB(),
            colors: [UIColor(hex: "#2e7d32").cgColor, UIColor(hex: "#1b5e20").cgColor] as CFArray,
            locations: [0, 1])!
        ctx.drawLinearGradient(grad,
            start: .zero, end: CGPoint(x: 0, y: L.headerHeight), options: [])

        // Score
        let scoreAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.boldSystemFont(ofSize: L.scoreFontSize),
            .foregroundColor: UIColor.white
        ]
        NSAttributedString(string: "Score: \(state.score)", attributes: scoreAttr)
            .draw(at: CGPoint(x: 16*L.scale, y: L.headerHeight/2 - L.scoreFontSize/2))

        // Best
        let bestAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.systemFont(ofSize: L.bestFontSize),
            .foregroundColor: UIColor(hex: "#a5d6a7")
        ]
        NSAttributedString(string: "Best: \(state.highScore)", attributes: bestAttr)
            .draw(at: CGPoint(x: 160*L.scale, y: L.headerHeight/2 - L.bestFontSize/2))

        // Volume button
        drawRoundedButton(ctx: ctx, rect: L.volumeBtn, fill: UIColor(hex: "#43a047"),
                          stroke: UIColor(hex: "#81c784"), label: "🔊", fontSize: 16*L.scale)

        // New Game button
        drawRoundedButton(ctx: ctx, rect: L.newGameBtn, fill: UIColor(hex: "#43a047"),
                          stroke: UIColor(hex: "#81c784"), label: "New Game", fontSize: L.buttonFontSize)
    }

    // MARK: - Game Over Overlay
    func drawGameOver(ctx: CGContext) {
        let L = layout
        // Dim grid
        ctx.setFillColor(UIColor(white: 0, alpha: 0.72).cgColor)
        ctx.fill(CGRect(x: 0, y: 0, width: L.gridWidth, height: L.totalHeight))

        // Panel
        let panelX: CGFloat = 40 * L.scale
        let panelW: CGFloat = L.gridWidth - 80 * L.scale
        ctx.setFillColor(UIColor(red: 0.106, green: 0.369, blue: 0.125, alpha: 0.97).cgColor)
        let panelPath = UIBezierPath(roundedRect: CGRect(x: panelX, y: 155*L.scale, width: panelW, height: 300*L.scale), cornerRadius: 16*L.scale)
        ctx.addPath(panelPath.cgPath); ctx.fillPath()
        ctx.setStrokeColor(UIColor(hex: "#4caf50").cgColor); ctx.setLineWidth(2)
        ctx.addPath(panelPath.cgPath); ctx.strokePath()

        let cx = L.gridWidth / 2

        // Title
        let titleAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.boldSystemFont(ofSize: L.gameOverTitleFont),
            .foregroundColor: UIColor.white
        ]
        let title = NSAttributedString(string: "Game Over", attributes: titleAttr)
        title.draw(at: CGPoint(x: cx - title.size().width/2, y: 215*L.scale - L.gameOverTitleFont/2))

        // Scores
        let scoreAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.systemFont(ofSize: L.gameOverBodyFont),
            .foregroundColor: UIColor(hex: "#a5d6a7")
        ]
        let s1 = NSAttributedString(string: "Final Score: \(state.score)", attributes: scoreAttr)
        s1.draw(at: CGPoint(x: cx - s1.size().width/2, y: 268*L.scale - L.gameOverBodyFont/2))

        let bestAttr2: [NSAttributedString.Key: Any] = [
            .font: UIFont.systemFont(ofSize: L.gameOverBodyFont),
            .foregroundColor: UIColor(hex: "#ffd54f")
        ]
        let s2 = NSAttributedString(string: "Best Score: \(state.highScore)", attributes: bestAttr2)
        s2.draw(at: CGPoint(x: cx - s2.size().width/2, y: 310*L.scale - L.gameOverBodyFont/2))

        // Buttons
        drawRoundedButton(ctx: ctx, rect: L.gobAdBtn, fill: UIColor(hex: "#2e7d32"),
                          stroke: UIColor(hex: "#81c784"), label: "▶ Watch Ad & Play Again",
                          fontSize: L.gameOverBtnFont * 0.82)
        drawRoundedButton(ctx: ctx, rect: L.gobBtn, fill: UIColor(hex: "#1b5e20"),
                          stroke: UIColor(hex: "#4caf50"), label: "Play Again",
                          fontSize: L.gameOverBtnFont * 0.88)
    }

    // MARK: - Helper
    func drawRoundedButton(ctx: CGContext, rect: CGRect, fill: UIColor, stroke: UIColor,
                            label: String, fontSize: CGFloat) {
        let path = UIBezierPath(roundedRect: rect, cornerRadius: 8 * layout.scale)
        ctx.setFillColor(fill.cgColor); ctx.addPath(path.cgPath); ctx.fillPath()
        ctx.setStrokeColor(stroke.cgColor); ctx.setLineWidth(1.5)
        ctx.addPath(path.cgPath); ctx.strokePath()

        let attr: [NSAttributedString.Key: Any] = [
            .font: UIFont.boldSystemFont(ofSize: fontSize),
            .foregroundColor: UIColor.white
        ]
        let str = NSAttributedString(string: label, attributes: attr)
        let sz = str.size()
        str.draw(at: CGPoint(x: rect.midX - sz.width/2, y: rect.midY - sz.height/2))
    }
}

File: android/app/src/main/java/com/flowerlines/rendering/GridRenderer.kt
package com.flowerlines.rendering

import android.graphics.*
import com.flowerlines.game.GameState
import com.flowerlines.game.GRID_SIZE
import com.flowerlines.layout.Layout

class GridRenderer(private val layout: Layout, private val state: GameState) {

    private val fillPaint   = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.STROKE }
    private val textPaint   = Paint(Paint.ANTI_ALIAS_FLAG).apply { textAlign = Paint.Align.CENTER }

    fun drawGrid(canvas: Canvas) {
        val L = layout
        // Green gradient
        val grad = LinearGradient(
            0f, L.headerHeight, 0f, L.headerHeight + L.cellSize * GRID_SIZE,
            Color.parseColor("#dcedc8"), Color.parseColor("#c5e1a5"), Shader.TileMode.CLAMP
        )
        fillPaint.shader = grad
        canvas.drawRect(0f, L.headerHeight, L.gridWidth, L.headerHeight + L.cellSize * GRID_SIZE, fillPaint)
        fillPaint.shader = null

        // Checkerboard + grid lines
        for (r in 0 until GRID_SIZE) {
            for (c in 0 until GRID_SIZE) {
                val x = c.toFloat() * L.cellSize
                val y = L.headerHeight + r.toFloat() * L.cellSize
                if ((r + c) % 2 == 0) {
                    fillPaint.color = Color.argb(56, 255, 255, 255)
                    canvas.drawRect(x, y, x + L.cellSize, y + L.cellSize, fillPaint)
                }
                strokePaint.color = Color.argb(140, 130, 180, 120)
                strokePaint.strokeWidth = 0.7f
                canvas.drawRect(x + 0.35f, y + 0.35f, x + L.cellSize - 0.7f, y + L.cellSize - 0.7f, strokePaint)
            }
        }
    }

    fun drawHighlights(canvas: Canvas) {
        val moves = state.validMoves ?: return
        val L = layout
        for (pos in moves) {
            val x = pos.col.toFloat() * L.cellSize + 1
            val y = L.headerHeight + pos.row.toFloat() * L.cellSize + 1
            fillPaint.color = Color.argb(102, 255, 230, 50)
            canvas.drawRect(x, y, x + L.cellSize - 2, y + L.cellSize - 2, fillPaint)
            fillPaint.color = Color.argb(64, 200, 170, 0)
            canvas.drawCircle(L.cellCX(pos.col), L.cellCY(pos.row), 4f, fillPaint)
        }
    }

    fun drawScoreBar(canvas: Canvas) {
        val L = layout
        val grad = LinearGradient(0f, 0f, 0f, L.headerHeight,
            Color.parseColor("#2e7d32"), Color.parseColor("#1b5e20"), Shader.TileMode.CLAMP)
        fillPaint.shader = grad
        canvas.drawRect(0f, 0f, L.gridWidth, L.headerHeight, fillPaint)
        fillPaint.shader = null

        val cy = L.headerHeight / 2f
        textPaint.apply {
            textAlign = Paint.Align.LEFT
            color = Color.WHITE
            textSize = L.scoreFontSize
            isFakeBoldText = true
        }
        canvas.drawText("Score: ${state.score}", 16f * L.scale, cy + L.scoreFontSize/3, textPaint)

        textPaint.apply { color = Color.parseColor("#a5d6a7"); textSize = L.bestFontSize; isFakeBoldText = false }
        canvas.drawText("Best: ${state.highScore}", 160f * L.scale, cy + L.bestFontSize/3, textPaint)

        // Volume button
        drawRoundedButton(canvas, L.volumeBtn, Color.parseColor("#43a047"),
            Color.parseColor("#81c784"), "🔊", 16f * L.scale, bold = false)

        // New Game button
        drawRoundedButton(canvas, L.newGameBtn, Color.parseColor("#43a047"),
            Color.parseColor("#81c784"), "New Game", L.buttonFontSize)
    }

    fun drawGameOver(canvas: Canvas) {
        val L = layout
        fillPaint.color = Color.argb(184, 0, 0, 0)
        canvas.drawRect(0f, 0f, L.gridWidth, L.totalHeight, fillPaint)

        val panelX = 40f * L.scale
        val panelW = L.gridWidth - 80f * L.scale
        val panelRect = RectF(panelX, 155f * L.scale, panelX + panelW, (155f + 300f) * L.scale)

        fillPaint.color = Color.argb(247, 27, 94, 32)
        canvas.drawRoundRect(panelRect, 16f * L.scale, 16f * L.scale, fillPaint)
        strokePaint.color = Color.parseColor("#4caf50"); strokePaint.strokeWidth = 2f
        canvas.drawRoundRect(panelRect, 16f * L.scale, 16f * L.scale, strokePaint)

        val cx = L.gridWidth / 2f
        textPaint.apply { textAlign = Paint.Align.CENTER; color = Color.WHITE; textSize = L.gameOverTitleFont; isFakeBoldText = true }
        canvas.drawText("Game Over", cx, 215f * L.scale + L.gameOverTitleFont / 3, textPaint)

        textPaint.apply { color = Color.parseColor("#a5d6a7"); textSize = L.gameOverBodyFont; isFakeBoldText = false }
        canvas.drawText("Final Score: ${state.score}", cx, 268f * L.scale + L.gameOverBodyFont / 3, textPaint)

        textPaint.color = Color.parseColor("#ffd54f")
        canvas.drawText("Best Score: ${state.highScore}", cx, 310f * L.scale + L.gameOverBodyFont / 3, textPaint)

        drawRoundedButton(canvas, L.gobAdBtn, Color.parseColor("#2e7d32"),
            Color.parseColor("#81c784"), "▶ Watch Ad & Play Again", L.gameOverBtnFont * 0.82f)
        drawRoundedButton(canvas, L.gobBtn, Color.parseColor("#1b5e20"),
            Color.parseColor("#4caf50"), "Play Again", L.gameOverBtnFont * 0.88f)
    }

    private fun drawRoundedButton(
        canvas: Canvas, rect: RectF, fill: Int, stroke: Int,
        label: String, fontSize: Float, bold: Boolean = true
    ) {
        val corner = 8f * layout.scale
        fillPaint.color = fill
        canvas.drawRoundRect(rect, corner, corner, fillPaint)
        strokePaint.color = stroke; strokePaint.strokeWidth = 1.5f
        canvas.drawRoundRect(rect, corner, corner, strokePaint)
        textPaint.apply { textAlign = Paint.Align.CENTER; color = Color.WHITE; textSize = fontSize; isFakeBoldText = bold }
        canvas.drawText(label, rect.centerX(), rect.centerY() + fontSize / 3, textPaint)
    }
}
